docs/contrib/DeveloperGuide.md

```diff
@@ -0,0 +1,352 @@
+# Developer Guide
+
+## Tool Philosophy
+
+OpenROAD is a tool to build a chip from a synthesized netlist to a
+physical design for manufacturing.
+
+The unifying principle behind the design of OpenROAD is for all of the
+tools to reside in one tool, with one process, and one database. All
+tools in the flow should use Tcl commands exclusively to control them
+instead of external "configuration files". File based communication
+between tools and forking processes is strongly discouraged. This
+architecture streamlines the construction of a flexible tool flow and
+minimizes the overhead of invoking each tool in the flow.
+
+## Tool File Organization
+
+Every tool follows the following file structure, grouping sources, tests
+and headers together.
+
+```
+    src/CMakelists.txt - add_subdirectory's src/CMakelists.txt
+    src/tool/src/ - sources and private headers
+    src/tool/src/CMakelists.txt
+    src/tool/include/tool/ - exported headers
+    src/tool/test/
+    src/tool/test/regression
+```
+
+OpenROAD repository
+
+```
+    CMakeLists.txt - top level cmake file
+    src/Main.cc
+    src/OpenROAD.cc - OpenROAD class functions
+    src/OpenROAD.i - top level swig, %includes tool swig files
+    src/OpenROAD.tcl - basic read/write lef/def/db commands
+    include/openroad/OpenRoad.hh - OpenROAD top level class, has instances of tools
+    include/openroad/Error.hh - Error reporting API
+```
+
+Some tools such as OpenDB are submodules, which are simply
+subdirectories in /src that are pointers to the git submodule. They are
+intentionally not segregated into a separate /module.
+
+The use of submodules for new code integrated into OpenROAD is strongly
+discouraged. Submodules make changes to the underlying infrastructure
+(OpenDB, OpenSTA etc) difficult to propagate across the dependent
+submodule repositories. Submodules: just say no.
+
+Where external/third party code that a tool depends on should be placed
+depends on the nature of the dependency.
+
+- Libraries - code packaged as a linkable library. Examples are tcl,
+    boost, zlib, eigen, lemon, spdlog.
+
+These should be installed in the build environment and linked by
+OpenRoad. Document these dependencies in the top level README.md file.
+The Dockerfile should be updated to illustrate where to find the library
+and how to install it. Adding libraries to the build enviroment requires
+coodination with the sys admins for the continuous integration hosts to
+make sure the environments include the dependency. Advanced notification
+should also be given to the development team so their private build
+environments can be updated.
+
```

Comment:
Document dependency handling reported by the issue: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues/106.

Reply:
Fixed on 6d5d6cd